### PR TITLE
Add run-as support for OAuth2 tokens

### PR DESCRIPTION
**Pull request by ywangd**
Authentication with an OAuth2 token can now perform run-as with changes
in this PR. This is in addition to the existing run-as support for realm
and API key authentication.

NB there are additional constraints on whether an OAuth2 token is
indeed qualified for run-as:
1. The token cannot itself already be a run-as (this can happen when the token is created using run-as and `client_credentials` grant)
2. The token cannot be created by anonymous or internal users

Service accounts cannot create tokens which meant run-as is not
supported for service accounts (none of existing service accounts
requires it anyway).

Relates: #84336

### CODE REVIEW

**Comment by justincr-elastic** [comment]
Can you clarify this in the description?
> The token cannot itself cannot already be a run-as

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/Authentication.java [hunk 1]
```diff
@@ -388,6 +388,55 @@ public boolean isApiKey() {
...
+        //
+        // This problem is at least partly due to we don't have special serialisation for the AnonymousUser class.
+        // As a result, it is serialised just as a normal user. At deserializing time, it is impossible to reliably
+        // tell the difference. This is what happens when AnonymousUser creates a token.
+        // Also, if anonymous access is disabled or anonymous username, roles are changed after the token is created.
+        // Should we still consider the token being created by an anonymous user which is now different from the new
+        // anonymous user?
+        if (getUser().equals(anonymousUser)) {
+            assert ANONYMOUS_REALM_TYPE.equals(getAuthenticatingSubject().getRealm().getType())
+                && ANONYMOUS_REALM_NAME.equals(getAuthenticatingSubject().getRealm().getName());
```
Comment (justincr-elastic) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/Authentication.java:426]: Should this be `||` based on your comment that a custom realm can return ANONYMOUS_REALM_NAME?

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/AuthenticatorChain.java [hunk 2]
```diff
@@ -203,15 +202,8 @@ void maybeLookupRunAsUser(Authenticator.Context context, Authentication authenti
...
-        // Both realm user and api_key can create tokens. They can also run-as another user and create tokens.
-        // In both cases, the created token will have a TOKEN authentication type and hence does not support run-as.
-        if (Authentication.AuthenticationType.REALM != authentication.getAuthenticationType()
-            && Authentication.AuthenticationType.API_KEY != authentication.getAuthenticationType()) {
-            logger.info(
-                "ignore run-as header since it is currently not supported for authentication type [{}]",
-                authentication.getAuthenticationType().name().toLowerCase(Locale.ROOT)
-            );
+        if (false == authentication.supportsRunAs(anonymousUser)) {
+            logger.info("ignore run-as header since it is currently not supported for authentication [{}]", authentication);
```
Comment (justincr-elastic) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/AuthenticatorChain.java:206]: Log parameter changed from `authentication.getAuthenticationType().name().toLowerCase(Locale.ROOT)` to `authentication`. Is that design intent?

**Comment by justincr-elastic** [comment]
I left a couple of drive by comments.

**Comment by justincr-elastic** [comment]
Are there any OAuth2 tests that use Access Tokens formatted as JWTs?

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/Authentication.java (hunk 1 above)
Comment (ywangd) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/Authentication.java:426]: The intention is to make sure we only treat ES's own anonymous user as the true anonymous user. So the user's realm must match both of the name and type of ES anonymous realm.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/AuthenticatorChain.java (hunk 2 above)
Comment (ywangd) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/AuthenticatorChain.java:206]: With the changes here, determine whether an authentication supports run-as is more involved and not just based on the authenticationType. So I changed it to log the whole authentication object, which will be easier for us to diagnose if the logging message ever comes to us in a SDH.

Technically we can log multiple different message but more precise about exactly which part of authentication preventing run-as. But I am not sure if it's worth it. Let me know if you feel strongly about it.

**Comment by ywangd** [comment]
> Are there any OAuth2 tests that use Access Tokens formatted as JWTs?

Not intentionally. But practically the `JwtRestIT` can be considered as one because for a JWT realm to work successfully, the ES OAuth2 token authenticator must have decided that it should yield from the authentication chain.

**Comment by lockewritesdocs** [comment]
@ywangd, as part of #86411, I'm clarifying that:

>Tokens that are created with the get token API are incompatible with the `run_as` privilege because OAuth2 tokens do not support this privilege.

This is true until your PR is merged. As part of this PR, I think that we should remove that sentence and then provide examples in the `run_as` section about how to generate OAuth2 tokens and use them with the `run_as` privilege.

**Comment by ywangd** [comment]
> As part of your PR, I think that we should remove that sentence and then provide examples in the `run_as` section about how to generate OAuth2 tokens and use them with the `run_as` privilege.

Makse sense. I'll get to it when your PR is merged. Thanks for the reminder!

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/Authentication.java [hunk 3]
```diff
@@ -388,6 +388,55 @@ public boolean isApiKey() {
         return effectiveSubject.getType() == Subject.Type.API_KEY;
     }
 
+    /**
+     * Whether the authentication can run-as another user
+     */
+    public boolean supportsRunAs(AnonymousUser anonymousUser) {
```
Comment (tvernum) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/Authentication.java:394]:
```suggestion
    public boolean supportsRunAs(@Nullable AnonymousUser anonymousUser) {
```

**Review by tvernum — APPROVED** [review]
LGTM

**Comment by ywangd** [comment]
@lockewritesdocs This PR is now ready to be merged. Just need  a round of CI to pass. In this case, would it make more sense for you to update your PR on the docs? I am happy to help out with providing technical details if needed. Thanks!

**Comment by lockewritesdocs** [comment]
@ywangd, go ahead and merge your PR and I'll remove the note about OAauth2. Please let me know if you have examples that we can use for `run_as`.

**Comment by ywangd** [comment]
> go ahead and merge your PR and I'll remove the note about OAauth2. Please let me know if you have examples that we can use for `run_as`.

Thanks. I'll message you for examples in a separate channel.

**Pull request merged**